File: LibsodiumSealedCryptoboxEncryptionString/LibsodiumSealedCryptoboxEncryptionString.py
import nacl.utils
from nacl.public import PrivateKey, SealedBox

# https://github.com/pyca/pynacl/
# PyNaCl version 1.4.0
# https://pynacl.readthedocs.io/en/latest/

def sealedCryptoboxEncryptionToBase64(publicKey, plaintext):
  pk = nacl.public.PublicKey(publicKey)
  box = SealedBox(pk)
  encrypted = box.encrypt(plaintext.encode("ascii"))
  return base64Encoding(encrypted)

def sealedCryptoboxDecryptionFromBase64(privateKey,  ciphertextDecryptionBase64):
  ciphertext = base64Decoding(ciphertextBase64)
  sk = nacl.public.PrivateKey(privateKey)
  box = SealedBox(sk)
  decryptedtext = box.decrypt(ciphertext)
  return decryptedtext.decode('utf-8')

def generateRandomNonce():
    return nacl.utils.random(24)

def base64Encoding(input):
    dataBase64 = nacl.encoding.Base64Encoder.encode(input)
    dataBase64P = dataBase64.decode("UTF-8")
    return dataBase64P

def base64Decoding(input):
    return nacl.encoding.Base64Encoder.decode(input.encode("ascii"))

def insert_colon(string, index):
    return string[:index] + ':' + string[index:]

def remove_colon(string):
    return string.replace(':', '')

print("Libsodium sealed crypto box hybrid string encryption")

plaintext = "The quick brown fox jumps over the lazy dog"
print("plaintext: " + plaintext)

# encryption
print("\n* * * encryption * * *")
print("all data are in Base64 encoding")

# for encryption you need the public key from other party (publicKeyB)
publicKeyBBase64 = "jVSuHjVH47PMbMaAxL5ziBS9/Z0HQK6TJLw9X3Jw6yg="
# convert the base64 encoded keys
publicKeyB = base64Decoding(publicKeyBBase64)
ciphertextBase64 = sealedCryptoboxEncryptionToBase64(publicKeyB, plaintext)
print("publicKeyB:  " + publicKeyBBase64)
print("ciphertext:  " + ciphertextBase64)
print("output is (Base64) ciphertext")

print("\n* * * decryption * * *");
# received ciphertext
ciphertextReceivedBase64 = ciphertextBase64;
# for decryption you need your private key (privateKeyB)
privateKeyBBase64 = "yNmXR5tfBXA/uZjanND+IYgGXlrFnrdUiUXesI4fOlM="
privateKeyB = nacl.encoding.Base64Encoder.decode(privateKeyBBase64)

decryptedtext = sealedCryptoboxDecryptionFromBase64(privateKeyB, ciphertextReceivedBase64);
print("ciphertext:  " + ciphertextReceivedBase64);
print("input is (Base64) ciphertext");
print("privateKeyB: " + privateKeyBBase64);
print("decrypt.text:" + decryptedtext);

File: LibsodiumSealedCryptoboxEncryptionString/LibsodiumSealedCryptoboxEncryptionString.go
package main

import (
  "golang.org/x/crypto/nacl/box"
	"crypto/rand"
  "encoding/base64"
	"fmt"
	"io"
)

func main() {

  fmt.Printf("Libsodium sealed crypto box hybrid string encryption\n")

  plaintext := string("The quick brown fox jumps over the lazy dog")
  fmt.Printf("plaintext: " + plaintext + "\n")

  fmt.Printf("\n* * * encryption * * *\n");
  fmt.Printf("all data are in Base64 encoding\n");
  // for encryption you need the public key from other party (publicKeyB)
  publicKeyBBase64 := "jVSuHjVH47PMbMaAxL5ziBS9/Z0HQK6TJLw9X3Jw6yg="
  // convert the base64 encoded keys
  publicKeyB := []byte(Base64Decoding(publicKeyBBase64))
  ciphertextBase64 := string(SealedCryptoboxEncryptionToBase64(publicKeyB, plaintext))
  fmt.Printf("publicKeyB:  " + publicKeyBBase64 + "\n")
  fmt.Printf("ciphertext: " + ciphertextBase64 + "\n");
  fmt.Printf("output is (Base64) ciphertext\n");

  fmt.Printf("\n* * * Decryption * * *\n");
  // for decryption you need your private key (privateKeyB)
  privateKeyBBase64 := "yNmXR5tfBXA/uZjanND+IYgGXlrFnrdUiUXesI4fOlM="
  // convert the base64 encoded keys
  privateKeyB := []byte(Base64Decoding(privateKeyBBase64))
  ciphertextReceivedBase64 := string(ciphertextBase64);
  fmt.Printf("ciphertext: " + ciphertextReceivedBase64 + "\n");
  fmt.Printf("input is (Base64) ciphertext\n");
  decryptedtext := string(SealedCryptoboxDecryptionFromBase64(privateKeyB, publicKeyB, ciphertextReceivedBase64))
  fmt.Printf("decryptedtext: " + decryptedtext + "\n")
}

func SealedCryptoboxEncryptionToBase64(publicKey []byte, data string)(string) {
  plaintext := []byte(data)
  // generate random noncevar nonce [24]byte
  var nonce []byte
	if _, err := io.ReadFull(rand.Reader, nonce[:]); err != nil {
		panic(err)
	}
  var publicKey32 [32]byte
	copy(publicKey32[:], publicKey)
  encrypted, error := box.SealAnonymous(nil, plaintext, &publicKey32, rand.Reader)
  if (error != nil) {
		panic(error)
	}
  return Base64Encoding(encrypted)
}

func SealedCryptoboxDecryptionFromBase64(privateKey []byte, publicKey []byte, ciphertextDecryptionBase64 string)(string) {
  ciphertext := []byte(Base64Decoding(ciphertextDecryptionBase64))
  var privateKey32 [32]byte
	copy(privateKey32[:], privateKey)
  var publicKey32 [32]byte
	copy(publicKey32[:], publicKey)
  decrypted, ok := box.OpenAnonymous(nil, ciphertext, &publicKey32, &privateKey32)
	if !ok {
		panic("decryption error")
	}
  return string(decrypted)
}

func Base64Encoding(input []byte)(string) {
  return base64.StdEncoding.EncodeToString(input)
}

func Base64Decoding(input string)([]byte) {
  data, err := base64.StdEncoding.DecodeString(input)
	if err != nil {
		return data
	}
  return data
}
